### ISSUE TITLE
[Bug]: A post causes a UI overflow on Tuba App

### ISSUE THREAD

**Issue by kop316** [GeopJr/Tuba#914]
### Describe the bug

Hello!

This post seems to cause an overflow of the UI of the Tuba App: https://akkoma.meows.gay/objects/b0dcc9c5-ece9-49eb-a83a-7c2b8b60c1bd

Screenshot:
![Screenshot_2024-04-17_20-40-09](https://github.com/GeopJr/Tuba/assets/16343055/c09bce5f-12af-4377-bc8f-45d7d533ae54)

### Steps To Reproduce

1. Follow that account
2. 2. Have it show up on your Home feed.

### Logs and/or Screenshots

```
<paste your logs here>
```

### Instance Backend

Mastodon

### Operating System

Debian Trixie

### Package

OS repositories

**Comment by GeopJr** [comment]
That's not it (or at least it's doesn't reproduce it here)

In general, the issue is with GtkLabel + ellipsize + wrap measuring the widget size incorrectly. GtkLabels are beyond broken under multiple containers, see this one for another example https://github.com/GeopJr/Tuba/issues/756#issuecomment-1912794500

It has been reported again #719 and again #632 and again #626 and many other times, but unless I get the post (or rather label) actually causing it, reproducibly, I can't really do anything, including creating a minimal reproduction so gtk fixes it or a workaround from our side

Finding which one is actually causing it is also somewhat difficult as you have to play around with the inspector until you figure it out, which is likely to crash a few times. I don't expect anyone to go through that but I'm still throwing it out there just in case

**Comment by GeopJr** [comment]
Though, from the looks of it you are also using an older version of Tuba, it might have been fixed on 0.7 :shrug:

**Comment by kop316** [comment]
Ahh, thanks for that. You are correct, that was the wrong post. I think I found the correct post that causes it: https://fosstodon.org/deck/@yassie_j@labyrinth.zone/112287918084026585

**Comment by kop316** [comment]
It looks like since this is a duplicate, it doesn't make sense to keep open.

**Comment by GeopJr** [comment]
Thanks for figuring it out!

I'll re-open this as they are mostly case-by-case. I'll see about writing a workaround!

**Comment by GeopJr** [comment]
Okay so, GtkLabel reports a larger min width that natural. This can be fixed by removing the overflow workaround but then we need to fix that. So for now, I'll say blocked by https://gitlab.gnome.org/GNOME/gtk/-/issues/6637 and see how it goes

**Comment by GeopJr** [comment]
I'm just going to post this here for future reference:

- Pango/GtkLabel issue
- Has to do with GtkLabel's minimum width being > natural width for height, only when there are custom shapes (/ custom emojis)
- Setting pango wrap to word from word_char seems to fix it, who knows what it broke however :shrug: 
- There are already way too many hacks around wrapping to avoid bugs such as https://gitlab.gnome.org/GNOME/gtk/-/issues/6637

**Comment by bugaevc** [comment]
Ahem... I still hope to eventually fix all of this from GTK side, once and for all. But evidently this is taking longer than expected, and I understand that you want to ship a non-broken app now (which I as a user only appreciate).

But please avoid doing anything (e.g.: designing LWW to fundamentally rely on hacks) that would prevent you from just doing the supported, straightforward thing in the future, once `GtkLabel` is fixed to just work. And please share any feedback (the more the better!) on what `GtkLabel` behavior/features would actually be useful for you. [This](https://github.com/GeopJr/Tuba/issues/756#issuecomment-2067579840) was the little piece of feedback that you shared last time, and that was very useful, and caused me to reevaluate how I want this to work.

**Comment by GeopJr** [comment]
Thanks, I appreciate all your work on fixing GtkLabel :bow:

As it stands, #1172 is not that big of a deal, it just changes wrapping from word_char to word. It might cause wrapping to not be as nice as it was but it's preferred over having your whole timeline broken just because of a single post

> And please share any feedback (the more the better!) on what `GtkLabel` behavior/features would actually be useful for you.

Well, the obvious one would be, adding custom widgets but... it's a very niche use that is more likely to make GtkLabel even more complex. So the second best solution, would probably be unsealing GtkLabel so I can freely manipulate its aria label and measuring, but I don't think anyone involved will be interested in that :sweat_smile: 

Other than that, if you go deep into pango stuff, `<bdi>` markup to isolate bi-directional characters would be useful #603, even for gnome-shell https://github.com/GeopJr/Tuba/issues/603#issuecomment-1762251451

And RTL languages are not always detected #16, I can't remember the specifics, this was ***very*** early in tuba's lifespan, but I ended up using a regex and set xalign manually

**Comment by bugaevc** [comment]
Dumping my findings before I forget: https://labyrinth.zone/notice/AgzGZoBtX749xlQGeG is Pango deciding to ellipsize around an inserted shape (for a custom emoji), which then [sets `item->num_chars`](https://gitlab.gnome.org/GNOME/pango/-/blob/main/pango/ellipsize.c#L658) to describe the whole ellipsized part, but then `pango_layout_run_get_extents_and_height ()` -> `_pango_shape_get_extents ()` [multiplies the shape's width by that `num_chars`](https://gitlab.gnome.org/GNOME/pango/-/blob/main/pango/pango-utils.c#L1129).

In other words: this is a Pango issue, related to the interaction between ellipsization and shapes (custom emoji). Matthias is the one who has enough understanding of Pango internals required to fix it; it would be great if you could make a simple reproducer that doesn't require all of Tuba — basically just LWW w/ that text — then we could submit the issue report to Pango with the above analysis.

**Comment by bugaevc** [comment]
I wrote a pure Pango reproducer myself: https://gitlab.gnome.org/GNOME/pango/-/issues/836

**Comment by bugaevc** [comment]
:tada: Totally fixed with https://gitlab.gnome.org/GNOME/pango/-/merge_requests/796 (in Pango main):

![Screenshot From 2024-12-27 16-41-03](https://github.com/user-attachments/assets/fe60111b-5efe-4ba2-8715-4054221c7723)
![Screenshot From 2024-12-27 16-40-08](https://github.com/user-attachments/assets/7d192793-0d92-4bd6-a6a6-5bd4cc882aa7)
![Screenshot From 2024-12-27 16-39-35](https://github.com/user-attachments/assets/55b8a382-3639-42af-bd2f-6d9a9556a7af)

And very likely wouldn't happen even w/o the Pango fix if you dropped the ellipsization hack due to improvements in GTK layouting.

**Comment by GeopJr** [comment]
Lets go!!!! Thank you so much :bow: I don't think anyone but you would have bothered pushing and fixing these :sweat_smile:

I notified Fractal some days ago about your work on LWW issues to see if they had any themselves; I think LWW is done for good :woozy_face: (they only had one but it was a race condition on their side, should be good now)
